ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -27,33 +54,164 @@ data Priority
+withLogMessageRecorder path action = do
+  textStderrRecorder <- threadSafeTextStderrRecorder
+  case path of
+    Nothing -> action (cmapIO messageToText textStderrRecorder)
+    Just path -> withTextFileRecorder path $ \textFileRecorder ->
+      action (cmapIO messageToText (textStderrRecorder <> textFileRecorder))
+
+makeDefaultStderrRecorder :: Priority -> IO (Recorder LogMessage)
+makeDefaultStderrRecorder minPriority = do
+  recorder <- threadSafeTextStderrRecorder
+  pure $ (cfilter (isLogMessagePriorityGreaterThan minPriority) . cmapIO messageToText) recorder
+  where
+    isLogMessagePriorityGreaterThan :: Priority -> LogMessage -> Bool
+    isLogMessagePriorityGreaterThan minPriority (LogOther LogOtherMessage{priority}) = priority >= minPriority
+
+logPriority :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Priority -> Text -> m ()
+logPriority Recorder{logger} priority text = withFrozenCallStack $
+  logger (LogOther LogOtherMessage{ stack = callStack, priority, payload = text })
+
+logException :: (HasCallStack, MonadIO m, Exception e) => Recorder LogMessage -> e -> m ()
+logException Recorder{logger} e = withFrozenCallStack $ logger message
+  where
+    message = LogOther LogOtherMessage{ stack = callStack, priority = Error, payload = Text.pack $ displayException e }
+
+logError :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Text -> m ()
+logError logger text = withFrozenCallStack $ logPriority logger Error text
+
+logWarning :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Text -> m ()
+logWarning logger text = withFrozenCallStack $ logPriority logger Warning text
+
+logInfo :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Text -> m ()
+logInfo logger text = withFrozenCallStack $ logPriority logger Info text
 
-instance Semigroup Logger where
-    l1 <> l2 = Logger $ \p t -> logPriority l1 p t >> logPriority l2 p t
+logDebug :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Text -> m ()
+logDebug logger text = withFrozenCallStack $ logPriority logger Debug text
 
-instance Monoid Logger where
-    mempty = Logger $ \_ _ -> pure ()
+logTelemetry :: (HasCallStack, MonadIO m) => Recorder LogMessage -> Text -> m ()
+logTelemetry logger text = withFrozenCallStack $ logPriority logger Telemetry text
 
-logError :: Logger -> T.Text -> IO ()
-logError x = logPriority x Error
+noopRecorder :: Recorder msg
+noopRecorder = mempty
 
-logWarning :: Logger -> T.Text -> IO ()
-logWarning x = logPriority x Warning
+noLogging :: Recorder msg
+noLogging = noopRecorder
 
-logInfo :: Logger -> T.Text -> IO ()
-logInfo x = logPriority x Info
+messageToText :: LogMessage -> IO Text
+messageToText = \case
+  LogOther LogOtherMessage{ stack, priority, payload } -> do
+    threadId <- myThreadId
+    utcTime <- getCurrentTime
+    pure $ Text.intercalate " | "
+      [ utcTimeToText utcTime
+      , callStackToLocationText stack
+      , threadIdToText threadId
+      , priorityToText priority
+      , payload ]
```

Comment:
I'm not keen on some of these details as I spend a lot of time reading test logs from CI:

- timestamps can be recorded by the CI system (or your terminal)\
- thread ids are not particularly helpful most of the time

Reply:
I think we can config it so when testing no timestamps or thread id, because I think timestamps can be useful for users reporting issues.

Reply:
For anything that uses the ghcide executable it should be very similar to how it is now.